### DISCUSSION TITLE
How to calculate final gear diameter

### DISCUSSION THREAD

**Discussion by x0pherl** [GarryBGoode/py_gearworks#8] [Q&A]
Hey, I just found this library when looking for a solution to building a herringbone gearchain and I have two questions both of which relate to understanding the final dimensions for the chain. I'm afraid that I'm not super familiar with involute gears, so some of these issues may stem from my own inexperience.

1) If I set the height of the HelicalGear to 10, the resulting part is 11.55mm high, and an axle cut in the center to the same "height" doesn't cut through. Am I misunderstanding the "height"? Are my other parameters forcing a change? Or is this a bug?
here's a minimal sample to show this issue:
```
gear_height = 10
with BuildPart() as gear:
    add(
        HelicalGear(
            number_of_teeth=22,
            module=1,
            height=gear_height,
            helix_angle=PI / 6,
            herringbone=True,
        ).build_part()
    )
    Cylinder(
        radius=4,
        height=gear_height,
        align=(Align.CENTER, Align.CENTER, Align.MIN),
        mode=Mode.SUBTRACT,
    )
```
2) How can I calculate or understand the distance required between two gears? Or better yet, how can I construct the right gear given a desired spacing distance of x?

**Comment by GarryBGoode** [comment]
Hi!

To the best of my knowledge, you're my first organic user, so... congrats!

1: You're not wrong, you found a bug. 

I'm not sure yet what it is, I can confirm that the herringbone option causes the height to be calculated badly, while with the usual helical gear appears get the height right. 

I'm looking into it. Could you confirm which version you used? Latest?

2: For spur gears, the gear's pitch circle diameter is defined as module\*number_of_teeth. Meshing gears should have tangential pitch-circles. So the distance should be the sum of the 2 radii, module\*(teeth1 + teeth2)/2. 

For helical gears, there is 1 complication with the transverse and normal system. The 'normal' system is straightforward from CAD perspective, a twist-extrude of the 2d gear profile, and the same equations apply. The 'transverse' system is the straightforward method from manufacturing perspective, they tilt the cutting tool while making them. But this tilting  will mean that the cutting dimension on the gear's plane is different - the pressure angle and the module changes.

There's a set of conversion equations that I'm planning to implement to help with helicals, for now I kind of hastily put together the transverse method for helicals - and I think I messed up somewhere with this part for the herringbone option.

If you're planning to make a planetary drive, I also have an example for it in 'examples.py'.

**Comment by x0pherl** [comment]
1. installed from a git pull: 0.0.3 (FYI it also looks like pip "requires" could be updated to include some stuff -- If you'd prefer me to break these points down into separate github issues for you, I'd be happy to do that.)
```
pip show gggears
Name: gggears
Version: 0.0.3
Summary: A gear geometry generator in Python
Home-page: 
Author: 
Author-email: Gergely Bencsik <bencsik.gergely.1@gmail.com>
License: 
Location: c:\Users\xopher\code\3d-print\house\blinds\.venv\Lib\site-packages
Requires: 
Required-by: 
```

2. Okay, yeah that explanation of what's going on with the helical gears helps a lot -- I kept comparing the generated STL to any math that I could think of and it just didn't reconcile. I would suggest that a helper property that calculates the appropriate distance between two gears would be very useful. For this design it probably doesn't matter a whole lot; I can sort of fudge it out based on physical prototyping (my first test is already printing) but I can how a that value would be incredibly useful for future designs.

**Reply by GarryBGoode** [reply]
Thank you!

I'm going to put together a To-Do topic to be transparent about what I'm working on.

I think I can make a quick-fix for your height problem. It can be done with a small change, but it's the symptom of a bigger issue with how I'm handling module... now that deserves an issue, I'll make one.

A question for pip: did it fail to add dependencies, did you have to manually install them?
I'm not that experienced with package creation, I probably missed this 'requires' field in the pyproject.toml, but I did fill the [dependencies] field.
I still need to add gggears to pypi, this topic will be part of that work package.

2...
Have you tried the `mesh_to()` function? You just make 2 gears and the `mesh_to()` function puts them in place for ideal distance.

Currently it's a bit convoluted, but you can get the coordinates of the 2 gears and then get the distance, something like this.

```
    gear1 = SpurGear(number_of_teeth=12, module=2, profile_shift=0.5)
    gear2 = SpurGear(
        number_of_teeth=24, module=2, enable_undercut=False, root_fillet=0.2
    )
    gear1.mesh_to(gear2, target_dir=UP)
    center1 = gear1.gearcore.transform.center
    center2 = gear2.gearcore.transform.center
    distance = np.linalg.norm(center1 - center2)

    gear_part_1 = gear1.build_part()
    gear_part_2 = gear2.build_part()
```

**Reply by x0pherl** [reply]
I'll test the height fix and update on the PR in the next day or so.

I already had build123d etc in my requirements.txt file so it installed just fine (I may submit a trivial docs PR with how to add gggears to requirements.txt). I'm not at all familiar with adding things to pip.

mesh_to worked great, but I'm building a frame for the axles and adding center cuts and stuff like that with GridLocations which is why I wanted the value. This workaround looks like a great solution for now

**Comment by x0pherl** [comment]
thanks for the quick fix; really happy to see this project taking shape, and I'm excited to see where it goes!
